CondeNast/launch-vehicle-fbm
# Postback check for 'greetings' & 'help' and payload consistency for postbacks and quick replies

## Why are we doing this?

Closes #38 

This PR ties up some loose ends on how postbacks are handled by checking the postback payload against the _greeting_ and _help_ regular expressions if enabled. Persistent menu clicks are emitted as postbacks and the SDK was not catching requests for _help_ that originated from a persistent menu.

Additionally, while `text` events normalized the user provided content by trimming whitespace and converting to all lower case; postbacks and quick replies were emitting payloads (as `text`) without this normalization. This PR emits the `text` in postback and quick replies as normalized text and adds a `payload` attribute that is the unmodified payload.

## Did you document your work?

Yes, `README` updated!

## How can someone test these changes?

Steps to manually verify the change:
1. `npm i`
2. `npm t`

Optionally, link to an existing bot and make sure said bot still works as expected:
1. `npm link`
2. `npm link launch-vehicle-fbm` wherever the test bot resides
3. run and exercise the test bot

## What possible risks or adverse effects are there?

An omission from the functionality is that the `text` events that are emitted from actual text (not payloads or quick replies) do not emit the unmodified text sent by the user, only the normalized text. Having this as the `payload` would be confusing as it would break the consistent association of `payload` with the unmodified content of a postback or quick reply. Thus, the `payload` attribute of the `text` event is not always sent.

## What are the follow-up tasks?

* address documentation on the choices we made in `message.postback` and `message.quickReply`, as it results from our developer experience involving how to handle text users type that mimics the `title` of a postback or quick reply; will likely do this after this PR and #55, which does some clean-up of the `README`, merge

## Are there any known issues?

None identified.

## Did the test coverage decrease?

No, new capability is covered.
